enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestionResult.tsx

```diff
@@ -1,80 +1,52 @@
 import cx from "classnames";
-import type React from "react";
-import { useUnmount } from "react-use";
 import { t } from "ttag";
 
-import type { SdkClickActionPluginsConfig } from "embedding-sdk";
 import {
   SdkError,
   SdkLoader,
 } from "embedding-sdk/components/private/PublicComponentWrapper";
 import { QuestionTitle } from "embedding-sdk/components/private/QuestionTitle";
 import { ResetButton } from "embedding-sdk/components/private/ResetButton";
-import { getDefaultVizHeight } from "embedding-sdk/lib/default-height";
-import { useSdkSelector } from "embedding-sdk/store";
-import { getPlugins } from "embedding-sdk/store/selectors";
+import { useInteractiveQuestionContext } from "embedding-sdk/components/public/InteractiveQuestion/context";
 import CS from "metabase/css/core/index.css";
-import { useDispatch, useSelector } from "metabase/lib/redux";
+import { useDispatch } from "metabase/lib/redux";
 import {
   navigateToNewCardInsideQB,
-  resetQB,
   updateQuestion,
 } from "metabase/query_builder/actions";
 import QueryVisualization from "metabase/query_builder/components/QueryVisualization";
 import { ViewHeaderContainer } from "metabase/query_builder/components/view/ViewHeader/ViewTitleHeader.styled";
 import {
-  DashboardBackButton,
   QuestionFiltersHeader,
+  DashboardBackButton,
 } from "metabase/query_builder/components/view/ViewHeader/components";
-import {
-  getCard,
-  getFirstQueryResult,
-  getQueryResults,
-  getQuestion,
-  getUiControls,
-} from "metabase/query_builder/selectors";
 import { Box, Group, Stack } from "metabase/ui";
-import { getEmbeddingMode } from "metabase/visualizations/click-actions/lib/modes";
-
-const returnNull = () => null;
 
 interface InteractiveQuestionResultProps {
-  isQuestionLoading: boolean;
-  onNavigateBack: () => void;
-  withResetButton?: boolean;
-  onResetButtonClick: () => void;
-  withTitle?: boolean;
-  customTitle?: React.ReactNode;
-  isOpenedFromDashboard?: boolean;
-
   height?: string | number;
-  componentPlugins?: SdkClickActionPluginsConfig;
 }
 
 export const InteractiveQuestionResult = ({
-  isQuestionLoading,
-  componentPlugins,
-  onNavigateBack,
   height,
-  withResetButton,
-  onResetButtonClick,
-  withTitle,
-  customTitle,
-  isOpenedFromDashboard = false,
 }: InteractiveQuestionResultProps): React.ReactElement => {
   const dispatch = useDispatch();
 
-  const globalPlugins = useSdkSelector(getPlugins);
-  const question = useSelector(getQuestion);
-  const card = useSelector(getCard);
-  const result = useSelector(getFirstQueryResult);
-  const uiControls = useSelector(getUiControls);
-  const queryResults = useSelector(getQueryResults);
-  const { isRunning: isQueryRunning } = uiControls;
-
-  useUnmount(() => {
-    dispatch(resetQB());
-  });
```

Comment:
We need to be careful with this change. Previously, we reset query builder every time the question result is unmounted, but now that we put this in the context, there's no guarantee how further up we can put this context. Technically, we can even put this in the app level, and it might work differently than how we imagine it to work.

This creates a quirk that could create more surface for new bugs. I'm not confident seeing this being moved to the context right now.

Reply:
I'm not sure I totally understand the explanation tbh. From what I understand, moving the this `reset` call to the provider would be the same as keeping it in the `InteractiveQuestionResult` component - if we mount the question and call `initializeQB`, it should be paired with an `unmount` with `resetQB`.

Alberto also mentioned that we might want to provide blank interactive question functionality in the future where we give the user the notebook _then_ display the `InteractiveQuestionResult`. If we put the `resetQB` in the question, if a user makes some changes to a notebook before the question is rendered, I don't think the store will reset, and if the user decides to create a new question their filters and aggregations will stay there

Let me know what you think - I don't have any strong opinions but I think in whatever case, `initialize` and `reset` should be paired in the same component with `mount` and `unmount`

Reply:
Because technically we can't guarantee where the provider is used. Take this absurd tree as an example

#### Previously
- App
  - Metabase provider
    - User Page
      - User component in a page
        - Interactive dashboard
          - interactive question result

Here QB is reset when the interactive question result is unmounted.

#### Currently (this is now how to use the context, but technically it could be done hence more surface for bugs)
- App
  - Metabase provider
    - The new interactive question provider
      - User Page
        - User component in a page
          - Interactive dashboard
            - interactive question result

Here QB won't be reset since the provider is at the page level despite the whole interactive dashboard being reset

My point is that we shouldn't couple this logic inside the provider causing the constraint that this context needs to be precisely positioned at a certain level.

I think context should be free to be placed anywhere without causing too much trouble. My mental model is that you can put context anywhere as long as the component using the context are within it.

Does that make sense?

Reply:
After huddling with Kelvin, we've come to an agreement that I will do two things to reduce the confusion about where to put this provider and prevent re-renders:

*   Move the redux selectors and their dependent data to a separate hook as they're not really needed within the context. Ideally, we'll be left with the display options (i.e. `withTitle`, `withResetButton`, etc)
*   Add a comment about the usage of the context within the code to ensure that it's only used within the right place.

Reply:
@WiNloSt let me know what you think - the commit is [here](https://github.com/metabase/metabase/pull/44730/commits/d6eb737194ea784b4a4e8d3b6f41d7dce7ee698b)